sessiontxn/staleread/util.go

```diff
@@ -65,3 +68,17 @@ func CalculateTsWithReadStaleness(sctx sessionctx.Context, readStaleness time.Du
 func IsStmtStaleness(sctx sessionctx.Context) bool {
 	return sctx.GetSessionVars().StmtCtx.IsStaleness
 }
+
+// GetExternalTimestamp returns the external timestamp in cache, or get and store it in cache
+func GetExternalTimestamp(ctx context.Context, sctx sessionctx.Context) (uint64, error) {
+	// Try to get from the stmt cache to make sure this function is deterministic.
+	stmtCtx := sctx.GetSessionVars().StmtCtx
+	externalTimestamp, err := stmtCtx.GetOrEvaluateStmtCache(stmtctx.StmtExternalTSCacheKey, func() (interface{}, error) {
+		return variable.GetExternalTimestamp(ctx)
+	})
+
+	if err != nil {
+		return 0, errAsOf.FastGenWithCause(err.Error())
+	}
+	return externalTimestamp.(uint64), nil
```

Comment:
no need to check the safe time now?

Reply:
Yes. It's the same behavior with staleness read variable. It's the same with `expression.CalAppropriateTime` with `minTime == maxTime == external timestamp` to make the behavior clearer.